### PR TITLE
Add code samples to provider page

### LINKED ISSUE #42: Idea: Show code samples like in RedwoodJS.com to reveal how to setup and use a provider
Idea: Show code samples like in RedwoodJS.com to reveal how to setup and use a provider

Similar to: 

image

This could be the actual code, but more like an example code snippet.

Maybe have a

image

similar to TailwindUI where preview is "try"? and code is the example?

### PR DESCRIPTION
**Pull request by morganmspencer**
# Overview

This PR adds a tab for code samples on the individual provider's page.

Closes #42 

- Adds toggle button
- Uses markdown files for the code
- Automatically pulls `.md` file if one exists
- Adds [highlight.js](https://highlightjs.org/) for styling
- Uses same highlight style as other Redwood sites
- Help wanted message for providers that do not have code

## To Do

- [x] Add code for each provider
- [x] Update CONTRIBUTING.md

## Screenshots

image

image

### CODE REVIEW

File: web/src/components/CodeSample/CodeSample.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import { useEffect, useState } from 'react'
+import hljs from 'highlight.js'
+import hljsDefineGraphQL from 'highlightjs-graphql'
+hljsDefineGraphQL(hljs)
+const md = require('markdown-it')({
```
Comment (Tobbe) [inline web/src/components/CodeSample/CodeSample.js:5]: I'm a big fan of remark, any reason you used markdown-it?

File: web/src/components/CodeSample/CodeSample.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    getMarkdown()
+  })
+
+  return (
+    <>
+      {markdownCode ? (
+        <div
+          className="markdown mt-6"
+          dangerouslySetInnerHTML={{ __html: markdownCode }}
```
Comment (Tobbe) [inline web/src/components/CodeSample/CodeSample.js:39]: With remark you don't have to use `dangerouslySetInnerHTML`

File: web/src/components/CodeSample/CodeSample.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    async function getMarkdown() {
+      try {
+        const file = await import(`../../lib/code-samples/${provider.slug}.md`)
+        setMarkdownCode(md.render(file.default))
+      } catch {
+        setMarkdownCode(null)
+      }
+    }
+    getMarkdown()
+  })
```
Comment (Tobbe) [inline web/src/components/CodeSample/CodeSample.js:32]:
```suggestion
  }, [provider.slug])
```

File: web/src/components/CodeSample/CodeSample.js (hunk 1 above)
Comment (morganmspencer) [reply web/src/components/CodeSample/CodeSample.js:5]: Markdown-it is used on redwoodjs.com so I actually just copied all the settings for it. Not opposed to switching to remark, just tried to maintain consistency

Comment (Tobbe) [reply web/src/components/CodeSample/CodeSample.js:5]: Thought so. Consistency is good. Keep it 👍 
Unless @dthyresson has a strong opinion

**Comment by morganmspencer** [comment]
@Tobbe and @dthyresson, for the code samples, should I include everything (AuthProvider, config, etc.) or just the utilization of the functions (logIn, logOut, etc.)?

It seems like having the first might not be necessary because that is all generated by RW.

**Comment by morganmspencer** [comment]
This is technically ready for review, I just couldn't quite wrap my head around adding the copy button to each potential `pre` block. That could be broken out into a separate issue.

Definitely needs checks on the code samples to make sure they are correct.

File: web/src/lib/code-samples/auth0.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        }}
+      >
+        {isAuthenticated ? 'Log Out' : 'Log In'}
+      </button>
+      {!isAuthenticated && (
+        <button
+          onClick={() => {
+            signUp()
+          }}
+        >
```
Comment (Tobbe) [inline web/src/lib/code-samples/auth0.md:21]:
```suggestion
        <button onClick={signUp}>
```

Just a code style preference from my side here. Let's see what @dthyresson says before changing anything

File: web/src/lib/code-samples/firebase.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+              signUp()
+            }}
+          >
+            Sign Up
+          </button>
+        )}
+      ) : (
+        <>
+          {isAuthenticated ? (
+            <button className="btn" onClick={() => logOut()}>
```
Comment (Tobbe) [inline web/src/lib/code-samples/firebase.md:56]:
```suggestion
            <button className="btn" onClick={logOut}>
```

**Comment by Tobbe** [comment]
> It seems like having the first might not be necessary because that is all generated by RW.

I agree. What you have now is good.

File: web/src/components/CodeSample/CodeSample.js (hunk 1 above)
Comment (dthyresson) [reply web/src/components/CodeSample/CodeSample.js:5]: Nope! Let's keep for now unless there's a feature in remark we need.

File: web/src/components/CodeSample/CodeSample.js (hunk 2 above)
Comment (dthyresson) [reply web/src/components/CodeSample/CodeSample.js:39]: Resolving since decided not to use remark for moment.

**Comment by dthyresson** [comment]
> I just couldn't quite wrap my head around adding the copy button to each potential `pre` block. That could be broken out into a separate issue.

@morganmspencer That sounds good. If you want to add that issue and tackle separately (or have otehrs work on it) that makes sense.

I accepted the deploy preview here: https://607e2f16e3f5160007178d99--redwood-playground-auth.netlify.app/

Only thing I noticed was that the Magic Link logo was missing:

image

**Comment by morganmspencer** [comment]
> Only thing I noticed was that the Magic Link logo was missing

My fault, I updated the provider images to just pull from the images folder if it exists. That way we don't need to explicitly define the provider image in `providers.js`, similar to the code samples. I think MagicLink and Azure were camelcased file names so it skipped them.

I've fixed this by adding a lowercase requirement for the image files in the latest commit.

**Comment by morganmspencer** [comment]
@dthyresson do you think I should add the RW setup commands above each code example as well?

**Comment by dthyresson** [comment]
> do you think I should add the RW setup commands above each code example as well?

I was wondering about that and also the client init (with the AuthProvider etc).

But perhaps it is best to just link to the documentation per provider:

* https://redwoodjs.com/docs/authentication#auth0
* https://redwoodjs.com/docs/authentication#supabase

etc

Perhaps the provider data has a "docsUrl" or something like that to store the link ... and some supporting wording: "How to setup ..."?

**Comment by morganmspencer** [comment]
Added the docsUrl for each provider and implemented it for the code samples

**Comment by dthyresson** [comment]
Deploy preview is up. LGTM. I'm merging.

**Pull request merged**